src/rand.jl

```diff
@@ -0,0 +1,61 @@
+
+import Random
+
+for (T, RefT) in ((Arf, ArfOrRef), (Arb, ArbOrRef), (Acb, AcbOrRef))
+    @eval begin
+        function Random.Sampler(
+            RNG::Type{<:Random.AbstractRNG},
+            st::Random.SamplerType{<:$RefT},
+            n::Random.Repetition,
+        )
+            #     TODO: SamplerBigFloat(precision(Arf))
+            #     return Random.SamplerSimple(
+            #         Random.SamplerType{Arf}(),
+            #         Random.SamplerBigFloat(precision(Arf))
+            #     )
+            res = setprecision(BigFloat, precision($T)) do
+                Random.SamplerSimple(Random.SamplerType{$T}(), Random.Sampler(RNG, BigFloat, n))
+            end
+            return res
+        end
+    end
+end
```

Comment:
If we had a method that given a ref-type returned the base-type then this could be written without the `@eval` right? I think that would make it slightly cleaner. 

I have also wanted a method like that in some other cases, in general converting between the base and ref type. We don't necessarily have to add it now though.

Reply:
yep, I could write something like
```julia
_nonreftype(::Type{ArfOrRef}) = Arf
_nonreftype(::Type{ArbOrRef}) = Arb
_nonreftype(::Type{AcbOrRef}) = Acb

function Random.Sampler(
    RNG::Type{<:Random.AbstractRNG},
    st::Random.SamplerType{TOrRef},
    n::Random.Repetition,
) where TOrRef <: Union{Arf, ArfRef, Arb, ArbRef, Acb, AcbRef}
    T = _nonreftype(TOrRef)
    # return Random.SamplerSimple(
    #     Random.SamplerType{T}(),
    #     Random.SamplerBigFloat(precision(T)),
    # )
    res = setprecision(BigFloat, precision(T)) do
        Random.SamplerSimple(Random.SamplerType{T}(), Random.Sampler(RNG, BigFloat, n))
    end
    return res
end
```
Should I go for it here?

btw. In the longer run I really think we should separate our code generation into `Meta` module, so it doesn't pollute `Arblib` namespace and prefix non-Arb helper functions by `_`...

Reply:
I would say go for it here! 

I do agree that it's annoying when all the actual code and all the meta code is mixed. Trying to have internal methods with an underscore is a good way to at least be able to tell them apart quickly. All the different type-unions we have defined also makes some things hard to read.